site/content/languages/zig.md

```diff
@@ -0,0 +1,213 @@
++++
+title = "Zig"
++++
+
+## Introduction
+
+Beginning with 0.4.0 [Zig][1] can generate `%.wasm` files instead of
+architecture-specific binaries through three targets:
+
+* `wasm32-emscripten`: mostly for browser (JavaScript) use.
+* `wasm32-freestanding`: for standalone use in or outside the browser.
+* `wasm32-wasi`: for use outside the browser.
+
+This document is maintained by wazero, which is a WebAssembly runtime that
+embeds in Go applications. Hence, our notes focus on targets used outside the
+browser, tested by wazero: `wasm32-freestanding` and `wasm32-wasi`.
+
+## Overview
+
+When Zig compiles a `%.zig` file with a `wasm32-*` target, the output `%.wasm`
+depends on a subset of features in the [WebAssembly 2.0 
+Core specification]({{< ref "/specs#core" >}}) and [WASI]({{< ref "/specs#wasi" >}}) host
+functions.
+
+Unlike some compilers, Zig also supports importing custom host functions and
+exporting functions back to the host.
+
+Here's a basic example of source in Zig:
+
+```zig
+export fn add(a: i32, b: i32) i32 {
+    return a + b;
+}
+```
+
+The following is the minimal command to build a Wasm file.
+
+```bash
+zig build-lib main.zig -target wasm32-freestanding -dynamic
+```
+
+The resulting Wasm `export`s the `add` function so that the embedding host can
+call it, regardless of if the host is written in Zig or not.
+
+Notice we are using `zig build-lib -dynamic`: this
+compiles the source as a library, i.e. without a `main` function.
+
+## Disclaimer
+
+This document includes notes contributed by the wazero community. While wazero
+includes Zig examples, and maintainers contribute to Zig, this
+isn't a Zig official document. For more help, consider the [Zig
+WebAssembly Guide](#???) or joining the [#programming-discussion channel on 
+Zig's Discord][5]. 
+
+Meanwhile, please help us [maintain][6] this document and [star our GitHub
+repository][7], if it is helpful. Together, we can make WebAssembly easier on
+the next person.
+
+## Constraints
+
+Please read our overview of WebAssembly and
+[constraints]({{< ref "_index.md#constraints" >}}). In short, expect
+limitations in both language features and library choices when developing your
+software.
+
+## Memory
+
+When Zig compiles a program into Wasm, it configures the WebAssembly linear memory
+to an initial size of 17 pages (1.1MB), and marks a position in that memory as
```

Comment:
> an initial size of 17 pages (1.1MB),

maybe add the permalink to this behavior would be helpful.

Reply:
Yeah I forgot to comment this line: TBH I just took a look at the generated `wasm` and I didn't find any references yet

Reply:
from further inspection. I think 17 (1088 KiB) may be because of [this behavior in LLVM's Wasm lld](https://lld.llvm.org/WebAssembly.html#cmdoption-initial-memory)

> **`--initial-memory=<value>`**
>     Initial size of the linear memory. Default: static data size.

even an "empty" WASI binary (nop `main`) contains a `data` section:

```
(data (i32.const 1048576) "InputOutput\00TruncatedInput...
```

i.e. 1024 KiB, which is about 17 pages

I don't know if it's worth documenting explicitly, though, it's basically a consequence of a consequence of a consequence :P

Reply:
I think this is dynamic and depends on the "data" of the compiled program - for example global variables are placed via data section's content and put into these memory. so tldr is that this size is not deterministic. so maybe we could just delete it.

Reply:
that's why this set by the linker - linker is the only component which knows the final size of data in the program.